pkg/syncmap/syncmap.go

```diff
@@ -0,0 +1,72 @@
+package syncmap
+
+import (
+	"sync"
+)
+
+// A SyncMap is a map of a string to a generified value which is locked for safe
+// access from multiple threads.
+// It is effectively a generic version of Golang's standard library sync.Map.
+type SyncMap[T any] struct {
+	data map[string]T
+	lock sync.Mutex
+}
+
+// NewSyncMap generates a new, empty SyncMap
+func New[T any]() *SyncMap[T] {
+	toReturn := new(SyncMap[T])
+	toReturn.data = make(map[string]T)
+
+	return toReturn
+}
+
+// Put adds an entry into the map
+func (m *SyncMap[T]) Put(key string, value T) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	m.data[key] = value
+}
+
+// Get retrieves an entry from the map.
+// Semantic match Golang map semantics - the bool represents whether the key
+// exists, and the empty value of T will be returned if the key does not exist.
+func (m *SyncMap[T]) Get(key string) (T, bool) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	value, exists := m.data[key]
+
+	return value, exists
+}
+
+// Exists returns true if a key exists in the map.
+func (m *SyncMap[T]) Exists(key string) bool {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	_, ok := m.data[key]
+
+	return ok
+}
+
+// Delete removes an entry from the map.
+func (m *SyncMap[T]) Delete(key string) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	delete(m.data, key)
+}
+
+// ToMap returns a shallow copy of the underlying data of the SyncMap.
```

Comment:
Do we want to actually clone that here? You did not clone them before so cloneing them adds some overhead, sure not enough to notice but still unnecessary. I guess that is my way of complainging that we cannot do this like you would in rust where that caller would have to take ownership of it and it would be clear that the syncmap is not used afterwards...

Reply:
I'll add a separate method to get the underlying map for perf reasons, with a caveat that doing so breaks all guarantees of thread safety. Doesn't matter for the way we're using it, where it is only ever accessed by a single thread after a certain point.